keps/sig-node/20200303-shared-pod-limits.md

```diff
@@ -0,0 +1,404 @@
+* NodeJS will call `madvise` if the memory reduction feature is used. Follow the code starting at https://github.com/nodejs/node/blob/master/deps/v8/src/heap/sweeper.cc#L322 
+
+The Java runtime autoconfigures some garbage collector sizes based on the amount of memory available to it. Since Java 9, the JDK is cgroup aware and will correctly use the amount of memory available to the cgroup that it is running in to perform this initialization. The OpenJDK implementation uses the `hierarchical_memory_limit` field in its `memory.stat` file to determine the amount of memory that is available to it. This field contains the maximum calculated amount of memory available to the cgroup based on the limits set on the entire memory controller hierarchy.
+
+When a process exits, all of the memory that was allocated to it is released back to the operating system and can be used by other processes. In some workloads, tasks are executed by forking and exec-ing child processes from some main process. In this case the memory being used by a short-lived process is made available to the rest of the pod immediately when the short-lived process exits.
+
+### User Stories 
+
+#### Story 1
+
+A development environment implemented as a Kubernetes Pod allows for separation of tools and a (web-based) IDE between multiple side-cars.
+
+The development environment defines a container with the web-server serving the IDE itself, and constrains it to use a certain amount of memory. Additional tools are provided in additional side-car deployments - for example an [LSP](https://langserver.org/) service, a terminal, and more. 
+
+Using this new feature, the containers providing the terminal, the LSP services, and the set of tools being utilized can share the resource limit defined for the pod. Consider the following pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: ide
+  name: myide
+spec:
+  resources:
+    limits:
+      memory: "1024M"
+      cpu: "4"
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: tool1
+    name: first-tool:latest
+  - image: tool2
+    name: second-tool:latest
+  - name: ide
+    image: theia:latest
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 
+
+Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.
+
+### Implementation Details/Notes/Constraints 
+
+The proposal is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should turn it on by specifying the relevant resource section in the Pod specification.
+
+### Interactions With Other Features
```

Comment:
Have you done any thinking around cri?  Currently cri only get container level resource requests/limits, and not pod-level. The pod level cgroups are controlled by the kubelet, but some cri-implementations like `kata` would probably need to get these pod-level limits in order to properly deal with them. Not 100% sure how this is handled today, so maybe someone else have any thoughts?

/cc @egernst

Reply:
According to [this](https://github.com/kata-containers/documentation/blob/master/how-to/how-to-set-sandbox-config-kata.md), `kata` allows setting annotations on the pod level so that pod-level information like `io.katacontainers.config.hypervisor.default_memory` can be configured on a per-pod basis. Otherwise, all of the containers belonging to a single pod are run inside the same OCI runtime (the `kata shimv2`, for example), and this process is placed inside the a single "outside" cgroup. Inside the `kata` VM, each pod is run in its own container. 

Implicitly, this means that if a specific container didn't have any limit set, it would be limited by the amount of memory allocated to the `kata` vm. If this is true - then no changes should be needed to support the `kata` use-case. If kubelet updates the pod-level cgroups in which the shim is being run, then everything should work correctly.   :-) I think :-)